python/ray/data/datasource/parquet_datasource.py

```diff
@@ -254,6 +280,53 @@ def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
 
         return read_tasks
 
+    def _estimate_files_encoding_ratio(self) -> float:
+        """Return an estimate of the Parquet files encoding ratio.
+
+        To avoid OOMs, it is safer to return an over-estimate than an underestimate.
+        """
+        if not DatasetContext.get_current().decoding_size_estimation:
+            return PARQUET_ENCODING_RATIO_ESTIMATE_DEFAULT
+
+        # Sample a few rows from Parquet files to estimate the encoding ratio.
+        # Launch tasks to sample multiple files remotely in parallel.
+        # Evenly distributed to sample N rows in i-th row group in i-th file.
+        # TODO(ekl/cheng) take into account column pruning.
+        start_time = time.perf_counter()
+        num_files = len(self._pq_ds.pieces)
+        num_samples = int(num_files * PARQUET_ENCODING_RATIO_ESTIMATE_SAMPLING_RATIO)
+        min_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MIN_NUM_SAMPLES, num_files
+        )
+        max_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MAX_NUM_SAMPLES, num_files
+        )
```

Comment:
We may want to truncate the max number of file samples based on the number of CPU cores in the cluster, in concert with decreasing the CPU request of the file sampling tasks (see comment below). cc @ericl

Reply:
It's truncated to 20 already, which should be much below the number of CPU cores. I'd rather the statistics not be dependent on the cluster size but only the input data (would be weird if the data size changed if you autoscaled up...)

Reply:
Yeah that makes sense, I'm mainly thinking of the getting started flow where a user is wanting to read some small Parquet files on a single machine, but I think the file sampling overhead should be negligible in that case anyway.